Idv form refactor

pkarman opened the pull request:
Refactors sessions controller into separate form+controller for each step.

**Why**: Server-side validation and better separation of concerns.

amoose commented on app/forms/idv/profile_form.rb:
```diff
@@ -0,0 +1,48 @@
+      Date.parse(dob)
+    rescue
+      errors.add :dob, I18n.t('idv.errors.bad_dob')
+    end
```
Should we also check for `< Date.today`?

amoose commented on config/locales/en.yml:
```diff
@@ -144,6 +144,8 @@ en:
       duplicate_ssn: >
         An account already exists using the information you provided.
+      missing_finance: You must provide a financial account ID.
+      bad_dob: Your date of birth must be a valid date.
```
Give feedback for valid format

pkarman replied:
The UI controls the format afaik. That's still in flux (see https://github.com/18F/identity-private/issues/701).

amoose replied:
👌

amoose commented on config/routes.rb:
```diff
@@ -68,15 +68,15 @@
+  get '/idv/phone' => 'idv/phone#new'
+  put '/idv/phone' => 'idv/phone#create'
+  get '/idv/review' => 'idv/review#new'
+  put '/idv/review' => 'idv/review#create'
```
I prefer explicit definitions in routes.rb 👍

sbc100 replied:
Maybe group all of these either before or after the `namespace :idv` block?

pkarman replied:
good call. I have made all the idv routes explicit, grouped together.

pkarman commented:
@monfresh would love your 👀  on this one since you suggested this pattern.

sbc100 commented on app/controllers/idv/finance_controller.rb:
```diff
@@ -0,0 +1,38 @@
+    private
+
+    def prep_vars
+      @idv_params = idv_params
```
I find the presence of `@idv_params` and `#idv_param` a little confusing here.   I'd also worry they could get out of sync.

Seems like `@idv_params` is just needed for the views, right?   Could you instead do `helper_method :idv_params` and remove the `@idv_params` completely?

sbc100 commented on spec/forms/idv/profile_form_spec.rb:
```diff
@@ -68,13 +70,30 @@
+        ).to eq false
+      end
+    end
+  end
```
These new tests seem unrelated?  Should they be a separate PR?

pkarman replied:
The dob tests were actually the impetus for this PR. They constitute the only new feature. I can pull all the additional validation checks into a separate PR if you like. That makes this PR a refactor only.

sbc100 replied:
Great.  Sorry if that seems nitpicky.

pkarman replied:
done in bbe30f46f96439208fbfc7dc289cefe3259a80ae

sbc100 commented:
Honestly I didn't review the entire thing because IIUC a lot of this is shuffling existing code around.

lgtm % comments

monfresh commented:
I'll take a look tomorrow. Been busy with the voice OTP PR today.

monfresh commented on app/controllers/idv/review_controller.rb:
```diff
@@ -0,0 +1,66 @@
+        redirect_on_success
+      else
+        flash[:error] = I18n.t('idv.titles.fail')
+        redirect_to idv_session_url
```
These two can be combined in a one-liner:
```ruby
redirect_to idv_session_url, alert: t('idv.titles.fail')
```
Note that by default, `redirect_to` only accepts certain types of flash keys, namely `alert` and `notice`, but Rails allows you to define additional flash types, with the `add_flash_types` method, which can be added to ApplicationController:
```ruby
add_flash_types :error
```

pkarman replied:
`error` seems to be used everywhere else in the app. Since there aren't any other instances in the app of this single line syntax, I'd prefer to keep it consistent. Good to know, though.

monfresh commented on app/controllers/idv/review_controller.rb:
```diff
@@ -0,0 +1,66 @@
+module Idv
+  class ReviewController < StepController
+    include IdvSession
```
Is this redundant, given that StepController includes it already?

pkarman replied:
yes good catch. missed that in my refactor.

monfresh commented on app/controllers/idv/review_controller.rb:
```diff
@@ -0,0 +1,66 @@
+    include IdvSession
+
+    before_action :confirm_two_factor_authenticated
+    before_action :confirm_idv_session_started
```
Same comment about these two being defined in StepController

monfresh commented on spec/controllers/idv/finance_controller_spec.rb:
```diff
@@ -0,0 +1,57 @@
+
+  describe '#create' do
+    before do
+      user = create(:user)
```
s/create(:user)/User.new. No need to use the DB for these tests, right?

monfresh commented on app/forms/idv/finance_form.rb:
```diff
@@ -0,0 +1,47 @@
+
+    def submit(params)
+      @finance_type = params[:finance_type]
+      @finance_account = params[:finance_account]
```
use `self.finance_type` and `self.finance_account` + a private `attr_writer :finance_type, :finance_account` for consistency with other forms?

monfresh commented on app/forms/idv/finance_form.rb:
```diff
@@ -0,0 +1,47 @@
+  class FinanceForm
+    include ActiveModel::Model
+
+    FINANCE_PARAMS = [:ccn, :mortgage, :home_equity_line, :auto_loan].freeze
```
rename this to `FINANCE_TYPES`?

pkarman replied:
good call. done in 7da3835

monfresh commented on app/forms/idv/finance_form.rb:
```diff
@@ -0,0 +1,46 @@
+      @finance_type = params[:finance_type]
+      @finance_account = params[:finance_account]
+      return false unless valid?
+      clear_idv_params_finance
```
Why is this necessary? Doesn't the next line achieve the same thing?

pkarman replied:
It doesn't. If you already have a financial type set (`:ccn`) and you change it (`:mortgage`) then you'll end up 2 unless we clear any existing first.

monfresh replied:
I see. Can we write a test for this?

How does the proofer gem parse this hash? What happens if there are 2 finance types?

monfresh commented on app/views/idv/finance/new.html.slim:
```diff
@@ -0,0 +1,15 @@
+    html: { autocomplete: 'off', method: 'put', role: 'form' }) do |f|
+  = f.error_notification
+  = f.collection_radio_buttons :finance_type, Idv::FinanceForm.finance_type_choices,
+    :first, :last, required: true, item_wrapper_tag: :div
```
Should we add make the first button selected by default? You can do that by adding `checked: Idv::FinanceForm.finance_type_choices.first`.

monfresh replied:
Also, it looks like we lost the styling we used to have, and the fact that each radio button had its own input field. I think we want separate input fields because the CCN field has to be exactly 8 digits, so I think we want to preserve the maxlength and pattern attributes we used to have.

pkarman replied:
The format/style of the finance form is still in flux. See https://github.com/18F/identity-private/issues/705#issuecomment-243254350 for example. The form *model* should be the same, though: a single text field. We'll have to apply client-side validation dynamically if we are going to enforce length or patterns. @brendansudol is aware of this.

monfresh replied:
So you're saying we're fine with losing functionality? We used to have pattern and length checks, but this PR removes them. To me, a refactoring should not change behavior.

pkarman replied:
I think we're fine for losing this small client-side validation check in the short term until the final UI is decided upon. I'd rather that we land this PR as-is and sort the validation in a separate PR with the other server-side validation checks.

monfresh replied:
Sounds good.

monfresh commented on app/forms/idv/finance_form.rb:
```diff
@@ -0,0 +1,46 @@
+    validates :finance_type, :finance_account, presence: true
+
+    validate :finance_type_valid
+
```
Add a validation for the CCN length?

pkarman replied:
That's a good idea. I'm trying to keep new features (like validation) out of this PR per @sbc100 request. I will add it in a separate PR.

monfresh commented on spec/controllers/idv/finance_controller_spec.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+
+    context 'when form is invalid' do
+      render_views
```
Rendering views in controller specs is expensive. I don't think we need to test every error message in a controller spec. I think it's enough to have a single test for the failure scenario that tests that the `new` action is rendered:
```ruby
expect(response).to render_template :new
```
I think feature tests are better suited for error messages, especially since we want to test the client-side error messages since most people will have JS enabled.

monfresh commented on spec/controllers/idv/review_controller_spec.rb:
```diff
@@ -0,0 +1,148 @@
+describe Idv::ReviewController do
+  render_views
+
+  let(:user) { create(:user, :signed_up, email: 'old_email@example.com') }
```
How about `User.new` to avoid using the DB?

sbc100 replied:
Isn't there a `create_stubbed` for that?

monfresh replied:
Yes, `build_stubbed`, but that uses the `factory_girl` gem. Why invoke more code than necessary? The User does not need to have any specific attributes for this test.

pkarman replied:
I have tried just `User.new` but because there is an access to `user_session` within the `create` method, Devise seems to want a real db user both signed in and present in the DB.

monfresh commented on spec/controllers/idv/review_controller_spec.rb:
```diff
@@ -0,0 +1,148 @@
+
+  describe '#new' do
+    before do
+      sign_in(user)
```
no need to sign in the user since you're stubbing the authentication below.

monfresh commented on app/controllers/concerns/idv_session.rb:
```diff
@@ -8,12 +8,24 @@ module IdvSession
+
+  def idv_phone_complete?
+    idv_params[:phone].present?
+  end
```
These 2 methods are only called in ReviewController. Move them there since they're not shared?

monfresh commented on spec/controllers/idv/review_controller_spec.rb:
```diff
@@ -0,0 +1,148 @@
+        get :new
+
+        user_attrs.values.each do |attr|
+          expect(response.body).to have_content(attr)
```
A view spec would be better to test this, as it is faster than a controller spec that renders views. Since the `new` action in this controller only symbolizes the keys in a hash, I would write a single test that makes sure the keys in the hash get symbolized. That is all the action is doing.

I would then write separate specs for the before_action methods to unit test them since we already have a spec that makes sure this controller includes those before_actions.

monfresh commented on spec/controllers/idv/review_controller_spec.rb:
```diff
@@ -0,0 +1,148 @@
+require 'rails_helper'
+
+describe Idv::ReviewController do
+  render_views
```
As mentioned before, calling `render_views` slows down tests, and it's only needed if you want to see text in the view, so it's rarely a good idea to call it at the top of a controller spec, because it unnecessarily slows down specs that don't test the view.

monfresh commented on spec/controllers/idv/review_controller_spec.rb:
```diff
@@ -0,0 +1,148 @@
+
+        expect(response).to redirect_to idv_finance_path
+      end
+    end
```
These 2 specs can be removed by unit testing the before_actions.

pkarman replied:
what would a unit test for the before_action look like that is different than these? I'm not seeing other examples in the spec/ dir for before_action-specific tests but perhaps I don't know what I'm looking for.

monfresh replied:
The duplication is that both the `new` and `create` specs are testing what happens when the user misses the phone and finance steps. Instead of repeating the test twice, we can have a single test that tests the `confirm_idv_steps_complete` method.

pkarman replied:
I've tried a couple different ways to test the `confirm_idv_steps_complete` method directly but am not having luck. Do you have a suggestion?

pkarman replied:
ok @monfresh I found a way in fdecba2

should be all ready for your re-review

monfresh commented on app/controllers/idv/sessions_controller.rb:
```diff
@@ -4,125 +4,41 @@ class SessionsController < ApplicationController
-      elsif @profile.errors.include?(:ssn)
+      if idv_profile_form.submit(profile_params)
+        redirect_to idv_finance_url
+      elsif idv_profile_form.errors.include?(:ssn)
```
There are multiple reasons the SSN could be invalid, but we only want to redirect if the error is related to the SSN not being unique. A blank SSN will trigger this as well. You can try it by submitting the form without entering anything (with JS disabled).

pkarman replied:
good catch. fix + test added.

monfresh commented on spec/forms/idv/finance_form_spec.rb:
```diff
@@ -0,0 +1,21 @@
+      expect(subject.submit(foo: 'bar')).to eq false
+    end
+  end
+end
```
Add specs to test the various validations? You can use shoulda_matchers for the presence validations.

monfresh commented on spec/forms/idv/profile_form_spec.rb:
```diff
@@ -59,7 +60,8 @@
     context 'when ssn is already taken by another profile' do
       it 'is invalid' do
-        create(:profile, ssn: '1234', user_id: 456)
+        diff_user = create(:user)
```
This can be tested without using the DB. See this test as an example: https://github.com/18F/identity-idp/blob/master/spec/forms/update_user_email_form_spec.rb#L23-L33

monfresh replied:
Although the stubbing might be more tricky here. Not a big deal if it's too complicated.

monfresh commented:
Done reviewing.

pkarman commented:
Thanks @monfresh I will take a look at refactoring some of the tests.

pkarman commented:
@monfresh I think I've addressed all your comments.

I'm reluctant at this point to optimize the tests themselves for DB access in ways that require any gymnastics in order to shave a second from the run time. I think I've avoided use of `create(:user)` where I can easily do that.

monfresh commented:
No problem about the test optimization if it's not straightforward. I'm planning to do a general sweep of tests to speed them up, so I'll tackle that later.

monfresh commented:
Looks good. Thanks!

monfresh commented:
Wanna squash or should I do it from the web?

pkarman commented:
I will squash.

pkarman commented:
all squashed. just waiting for the 💚

monfresh commented:
I checked the branch directly on Hakiri.io, and it passed, so I'm going to disable the Hakiri requirement temporarily so I can merge this.
